File: ios/LLMHub/Sources/LLMHub/MimoBot/Kokoro/EnglishG2P.swift
import Foundation

/// Tiny English grapheme-to-phoneme converter.
///
/// **Starter, not finished.** Hand-curated dictionary covers what a voice
/// companion is most likely to say (greetings, pronouns, numbers, common
/// verbs / nouns / adjectives). Out-of-vocabulary words fall back to
/// letter-by-letter spelling — usable but robotic.
///
/// Replace with espeak-ng or misaki for production. Drop-in shape is the
/// same: text → space-joined IPA using only `KokoroVocab.symbols`.
///
/// TODO(g2p):
///   1. Add an espeak-ng XCFramework (or a Swift port of misaki).
///   2. Replace `phonemize` with the upstream G2P call.
///   3. Match the exact phonemizer the chosen Kokoro ONNX export was
///      trained against (kokoro-onnx → espeak-ng with `--ipa=3`).
enum EnglishG2P {

    static func phonemize(_ text: String) -> String {
        let words = text.lowercased()
            .components(separatedBy: CharacterSet.letters.inverted.subtracting(["'"]))
            .filter { !$0.isEmpty }
        var parts: [String] = []
        for w in words {
            parts.append(dict[w] ?? spell(w))
        }
        return parts.joined(separator: " ")
    }

    private static func spell(_ word: String) -> String {
        word.compactMap { letter[$0] }.joined(separator: " ")
    }

    private static let letter: [Character: String] = [
        "a": "eɪ", "b": "bi", "c": "si", "d": "di", "e": "i",
        "f": "ɛf", "g": "ʤi", "h": "eɪʧ", "i": "aɪ", "j": "ʤeɪ",
        "k": "keɪ", "l": "ɛl", "m": "ɛm", "n": "ɛn", "o": "oʊ",
        "p": "pi", "q": "kju", "r": "ɑɹ", "s": "ɛs", "t": "ti",
        "u": "ju", "v": "vi", "w": "ˈdʌbəlju", "x": "ɛks", "y": "waɪ",
        "z": "zi"
    ]

    /// Hand-curated dictionary. Stress markers omitted; diphthongs split into
    /// adjacent vowel symbols since Kokoro tokenises per IPA char.
    private static let dict: [String: String] = [
        // greetings
        "hello": "həloʊ", "hi": "haɪ", "hey": "heɪ",
        "bye": "baɪ", "goodbye": "ɡʊdbaɪ",
        // affirm / deny
        "yes": "jɛs", "yeah": "jɛə", "yep": "jɛp",
        "no": "noʊ", "nope": "noʊp", "not": "nɑt",
        "maybe": "meɪbi", "ok": "oʊkeɪ", "okay": "oʊkeɪ",
        "sure": "ʃʊɹ", "alright": "ɔlɹaɪt",
        // pronouns
        "i": "aɪ", "me": "mi", "my": "maɪ", "mine": "maɪn",
        "you": "ju", "your": "jɔɹ", "yours": "jɔɹz",
        "we": "wi", "us": "ʌs", "our": "aʊɹ", "ours": "aʊɹz",
        "he": "hi", "him": "hɪm", "his": "hɪz",
        "she": "ʃi", "her": "hɚ", "hers": "hɚz",
        "it": "ɪt", "its": "ɪts",
        "they": "ðeɪ", "them": "ðɛm", "their": "ðɛɹ",
        // copula / aux
        "is": "ɪz", "am": "æm", "are": "ɑɹ",
        "was": "wʌz", "were": "wɚ", "be": "bi", "been": "bɪn",
        "have": "hæv", "has": "hæz", "had": "hæd",
        "do": "du", "does": "dʌz", "did": "dɪd",
        "will": "wɪl", "would": "wʊd", "can": "kæn", "could": "kʊd",
        "should": "ʃʊd", "shall": "ʃæl", "may": "meɪ", "might": "maɪt",
        // conjunctions / preps
        "and": "ænd", "or": "ɔɹ", "but": "bʌt", "if": "ɪf",
        "then": "ðɛn", "than": "ðæn", "so": "soʊ", "because": "bɪkʌz",
        "the": "ðə", "a": "ə", "an": "ən", "of": "əv",
        "in": "ɪn", "on": "ɔn", "at": "æt", "to": "tu",
        "from": "fɹʌm", "for": "fɔɹ", "with": "wɪð",
        "by": "baɪ", "about": "əbaʊt", "as": "æz",
        // questions
        "what": "wʌt", "who": "hu", "where": "wɛɹ", "when": "wɛn",
        "why": "waɪ", "how": "haʊ", "which": "wɪʧ",
        // demonstratives / location
        "this": "ðɪs", "that": "ðæt", "these": "ðiz", "those": "ðoʊz",
        "here": "hɪɹ", "there": "ðɛɹ", "now": "naʊ",
        // numbers
        "zero": "zɪɹoʊ", "one": "wʌn", "two": "tu", "three": "θɹi",
        "four": "fɔɹ", "five": "faɪv", "six": "sɪks", "seven": "sɛvən",
        "eight": "eɪt", "nine": "naɪn", "ten": "tɛn",
        // courtesy
        "please": "pliz", "thanks": "θæŋks", "thank": "θæŋk",
        "sorry": "sɑɹi", "excuse": "ɪkskjuz",
        // common verbs
        "go": "ɡoʊ", "going": "ɡoʊɪŋ", "gone": "ɡɔn",
        "come": "kʌm", "coming": "kʌmɪŋ",
        "see": "si", "saw": "sɔ", "seen": "sin",
        "say": "seɪ", "said": "sɛd", "tell": "tɛl", "told": "toʊld",
        "think": "θɪŋk", "thought": "θɔt",
        "know": "noʊ", "knew": "nu",
        "like": "laɪk", "want": "wɑnt", "wanted": "wɑntɪd",
        "need": "nid", "needs": "nidz",
        "help": "hɛlp", "make": "meɪk", "made": "meɪd",
        "take": "teɪk", "took": "tʊk", "taken": "teɪkən",
        "give": "ɡɪv", "gave": "ɡeɪv", "got": "ɡɑt", "get": "ɡɛt",
        "find": "faɪnd", "found": "faʊnd",
        "look": "lʊk", "ask": "æsk", "play": "pleɪ",
        "work": "wɝk", "open": "oʊpən", "close": "kloʊz",
        "start": "stɑɹt", "stop": "stɑp",
        // common nouns
        "thing": "θɪŋ", "things": "θɪŋz", "time": "taɪm",
        "day": "deɪ", "night": "naɪt", "today": "tədeɪ",
        "tomorrow": "təmɑɹoʊ", "yesterday": "jɛstɚdeɪ",
        "name": "neɪm", "person": "pɝsən", "people": "pipəl",
        "way": "weɪ", "world": "wɝld",
        "home": "hoʊm", "house": "haʊs",
        "music": "mjuzɪk", "song": "sɔŋ", "light": "laɪt",
        "lights": "laɪts", "volume": "vɑljəm", "battery": "bætəɹi",
        "robot": "ɹoʊbɑt", "bot": "bɑt", "voice": "vɔɪs",
        "model": "mɑdəl", "phone": "foʊn",
        // common adjectives
        "good": "ɡʊd", "bad": "bæd", "great": "ɡɹeɪt", "nice": "naɪs",
        "hot": "hɑt", "cold": "koʊld", "warm": "wɔɹm",
        "big": "bɪɡ", "small": "smɔl", "little": "lɪtəl",
        "old": "oʊld", "new": "nu", "young": "jʌŋ",
        "happy": "hæpi", "sad": "sæd", "fine": "faɪn",
        "right": "ɹaɪt", "wrong": "ɹɔŋ", "real": "ɹiəl",
        "slow": "sloʊ", "fast": "fæst", "loud": "laʊd", "quiet": "kwaɪət",
        // adverbs / common
        "very": "vɛɹi", "really": "ɹiəli", "too": "tu",
        "also": "ɔlsoʊ", "just": "ʤʌst", "only": "oʊnli",
        "more": "mɔɹ", "less": "lɛs", "most": "moʊst",
        "all": "ɔl", "any": "ɛni", "some": "sʌm", "none": "nʌn",
        "always": "ɔlweɪz", "never": "nɛvɚ", "again": "əɡɛn",
        // contractions
        "i'm": "aɪm", "you're": "jʊɹ", "he's": "hiz", "she's": "ʃiz",
        "it's": "ɪts", "we're": "wɪɹ", "they're": "ðɛɹ",
        "don't": "doʊnt", "doesn't": "dʌzənt", "didn't": "dɪdənt",
        "can't": "kænt", "won't": "woʊnt", "isn't": "ɪzənt",
        "aren't": "ɑɹənt", "wasn't": "wʌzənt",
        "i'll": "aɪl", "you'll": "jul", "we'll": "wil",
        "i've": "aɪv", "you've": "juv", "we've": "wiv",
        // commands a voice bot hears a lot
        "turn": "tɝn", "set": "sɛt", "show": "ʃoʊ",
        "pause": "pɔz",
        "next": "nɛkst", "back": "bæk", "off": "ɔf"
    ]
}

File: android/app/src/main/java/com/llmhub/llmhub/mimobot/speech/kokoro/EnglishG2P.kt
package com.llmhub.llmhub.mimobot.speech.kokoro

/**
 * Tiny English grapheme-to-phoneme converter.
 *
 * **This is a starter, not a finished G2P.** It ships a small hand-curated
 * dictionary covering the vocabulary a voice companion is most likely to need
 * (greetings, pronouns, numbers, common verbs/nouns/adjectives) so the Kokoro
 * pipeline produces real, intelligible audio for those words. Out-of-vocabulary
 * words fall back to a deterministic letter-by-letter spelling — usable but
 * sounds like a robot literally spelling things.
 *
 * Replace this with a proper espeak-ng or misaki-style phonemizer for
 * production. The expected drop-in shape is the same: text in → space-joined
 * IPA out, using only symbols present in [KokoroVocab.SYMBOLS].
 *
 * TODO(g2p):
 *   1. Compile espeak-ng for arm64-v8a + armeabi-v7a → libespeak-ng.so.
 *   2. Bundle espeak-ng-data (~7 MB) under src/main/assets/espeak-data.
 *   3. Replace [phonemize] with an espeak_TextToPhonemes JNI call.
 *   4. Match the exact phonemizer that the chosen Kokoro ONNX export was
 *      trained against (kokoro-onnx uses espeak-ng with `--ipa=3` flags).
 */
object EnglishG2P {

    /**
     * Convert a text utterance to a phoneme string ready for [KokoroVocab.tokenise].
     */
    fun phonemize(text: String): String {
        val tokens = text.lowercase().split(Regex("[^a-z']+")).filter { it.isNotEmpty() }
        val sb = StringBuilder()
        for ((i, w) in tokens.withIndex()) {
            if (i > 0) sb.append(' ')
            sb.append(DICT[w] ?: spell(w))
        }
        return sb.toString()
    }

    /** Last-resort fallback: speak the word letter by letter. */
    private fun spell(word: String): String {
        val sb = StringBuilder()
        for ((i, c) in word.withIndex()) {
            if (i > 0) sb.append(' ')
            sb.append(LETTER[c] ?: "")
        }
        return sb.toString()
    }

    /** ASCII letter → English letter name in IPA. */
    private val LETTER: Map<Char, String> = mapOf(
        'a' to "eɪ", 'b' to "bi", 'c' to "si", 'd' to "di", 'e' to "i",
        'f' to "ɛf", 'g' to "ʤi", 'h' to "eɪʧ", 'i' to "aɪ", 'j' to "ʤeɪ",
        'k' to "keɪ", 'l' to "ɛl", 'm' to "ɛm", 'n' to "ɛn", 'o' to "oʊ",
        'p' to "pi", 'q' to "kju", 'r' to "ɑɹ", 's' to "ɛs", 't' to "ti",
        'u' to "ju", 'v' to "vi", 'w' to "ˈdʌbəlju", 'x' to "ɛks", 'y' to "waɪ",
        'z' to "zi"
    )

    /**
     * Hand-curated starter dictionary. Stress markers are omitted for
     * simplicity — Kokoro handles bare phonemes OK. Diphthongs are written as
     * two adjacent vowel symbols (e.g. "oʊ", "aɪ") since Kokoro tokenises per
     * symbol.
     */
    private val DICT: Map<String, String> = mapOf(
        // greetings
        "hello" to "həloʊ", "hi" to "haɪ", "hey" to "heɪ",
        "bye" to "baɪ", "goodbye" to "ɡʊdbaɪ",
        // affirmation / negation
        "yes" to "jɛs", "yeah" to "jɛə", "yep" to "jɛp",
        "no" to "noʊ", "nope" to "noʊp", "not" to "nɑt",
        "maybe" to "meɪbi", "ok" to "oʊkeɪ", "okay" to "oʊkeɪ",
        "sure" to "ʃʊɹ", "alright" to "ɔlɹaɪt",
        // pronouns
        "i" to "aɪ", "me" to "mi", "my" to "maɪ", "mine" to "maɪn",
        "you" to "ju", "your" to "jɔɹ", "yours" to "jɔɹz",
        "we" to "wi", "us" to "ʌs", "our" to "aʊɹ", "ours" to "aʊɹz",
        "he" to "hi", "him" to "hɪm", "his" to "hɪz",
        "she" to "ʃi", "her" to "hɚ", "hers" to "hɚz",
        "it" to "ɪt", "its" to "ɪts",
        "they" to "ðeɪ", "them" to "ðɛm", "their" to "ðɛɹ",
        // copula / aux
        "is" to "ɪz", "am" to "æm", "are" to "ɑɹ",
        "was" to "wʌz", "were" to "wɚ", "be" to "bi", "been" to "bɪn",
        "have" to "hæv", "has" to "hæz", "had" to "hæd",
        "do" to "du", "does" to "dʌz", "did" to "dɪd",
        "will" to "wɪl", "would" to "wʊd", "can" to "kæn", "could" to "kʊd",
        "should" to "ʃʊd", "shall" to "ʃæl", "may" to "meɪ", "might" to "maɪt",
        // conjunctions / preps
        "and" to "ænd", "or" to "ɔɹ", "but" to "bʌt", "if" to "ɪf",
        "then" to "ðɛn", "than" to "ðæn", "so" to "soʊ", "because" to "bɪkʌz",
        "the" to "ðə", "a" to "ə", "an" to "ən", "of" to "əv",
        "in" to "ɪn", "on" to "ɔn", "at" to "æt", "to" to "tu",
        "from" to "fɹʌm", "for" to "fɔɹ", "with" to "wɪð",
        "by" to "baɪ", "about" to "əbaʊt", "as" to "æz",
        // questions
        "what" to "wʌt", "who" to "hu", "where" to "wɛɹ", "when" to "wɛn",
        "why" to "waɪ", "how" to "haʊ", "which" to "wɪʧ",
        // demonstratives / location
        "this" to "ðɪs", "that" to "ðæt", "these" to "ðiz", "those" to "ðoʊz",
        "here" to "hɪɹ", "there" to "ðɛɹ", "now" to "naʊ",
        // numbers
        "zero" to "zɪɹoʊ", "one" to "wʌn", "two" to "tu", "three" to "θɹi",
        "four" to "fɔɹ", "five" to "faɪv", "six" to "sɪks", "seven" to "sɛvən",
        "eight" to "eɪt", "nine" to "naɪn", "ten" to "tɛn",
        // courtesy
        "please" to "pliz", "thanks" to "θæŋks", "thank" to "θæŋk",
        "sorry" to "sɑɹi", "excuse" to "ɪkskjuz",
        // common verbs
        "go" to "ɡoʊ", "going" to "ɡoʊɪŋ", "gone" to "ɡɔn",
        "come" to "kʌm", "coming" to "kʌmɪŋ",
        "see" to "si", "saw" to "sɔ", "seen" to "sin",
        "say" to "seɪ", "said" to "sɛd", "tell" to "tɛl", "told" to "toʊld",
        "think" to "θɪŋk", "thought" to "θɔt",
        "know" to "noʊ", "knew" to "nu",
        "like" to "laɪk", "want" to "wɑnt", "wanted" to "wɑntɪd",
        "need" to "nid", "needs" to "nidz",
        "help" to "hɛlp", "make" to "meɪk", "made" to "meɪd",
        "take" to "teɪk", "took" to "tʊk", "taken" to "teɪkən",
        "give" to "ɡɪv", "gave" to "ɡeɪv", "got" to "ɡɑt", "get" to "ɡɛt",
        "find" to "faɪnd", "found" to "faʊnd",
        "look" to "lʊk", "ask" to "æsk", "play" to "pleɪ",
        "work" to "wɝk", "open" to "oʊpən", "close" to "kloʊz",
        "start" to "stɑɹt", "stop" to "stɑp",
        "is" to "ɪz", "do" to "du",
        // common nouns
        "thing" to "θɪŋ", "things" to "θɪŋz", "time" to "taɪm",
        "day" to "deɪ", "night" to "naɪt", "today" to "tədeɪ",
        "tomorrow" to "təmɑɹoʊ", "yesterday" to "jɛstɚdeɪ",
        "name" to "neɪm", "person" to "pɝsən", "people" to "pipəl",
        "thing" to "θɪŋ", "way" to "weɪ", "world" to "wɝld",
        "home" to "hoʊm", "house" to "haʊs",
        "music" to "mjuzɪk", "song" to "sɔŋ", "light" to "laɪt",
        "lights" to "laɪts", "volume" to "vɑljəm", "battery" to "bætəɹi",
        "robot" to "ɹoʊbɑt", "bot" to "bɑt", "voice" to "vɔɪs",
        "model" to "mɑdəl", "phone" to "foʊn",
        // common adjectives
        "good" to "ɡʊd", "bad" to "bæd", "great" to "ɡɹeɪt", "nice" to "naɪs",
        "hot" to "hɑt", "cold" to "koʊld", "warm" to "wɔɹm",
        "big" to "bɪɡ", "small" to "smɔl", "little" to "lɪtəl",
        "old" to "oʊld", "new" to "nu", "young" to "jʌŋ",
        "happy" to "hæpi", "sad" to "sæd", "fine" to "faɪn",
        "right" to "ɹaɪt", "wrong" to "ɹɔŋ", "real" to "ɹiəl",
        "slow" to "sloʊ", "fast" to "fæst", "loud" to "laʊd", "quiet" to "kwaɪət",
        // adverbs / common
        "very" to "vɛɹi", "really" to "ɹiəli", "too" to "tu",
        "also" to "ɔlsoʊ", "just" to "ʤʌst", "only" to "oʊnli",
        "more" to "mɔɹ", "less" to "lɛs", "most" to "moʊst",
        "all" to "ɔl", "any" to "ɛni", "some" to "sʌm", "none" to "nʌn",
        "always" to "ɔlweɪz", "never" to "nɛvɚ", "again" to "əɡɛn",
        // contractions
        "i'm" to "aɪm", "you're" to "jʊɹ", "he's" to "hiz", "she's" to "ʃiz",
        "it's" to "ɪts", "we're" to "wɪɹ", "they're" to "ðɛɹ",
        "don't" to "doʊnt", "doesn't" to "dʌzənt", "didn't" to "dɪdənt",
        "can't" to "kænt", "won't" to "woʊnt", "isn't" to "ɪzənt",
        "aren't" to "ɑɹənt", "wasn't" to "wʌzənt",
        "i'll" to "aɪl", "you'll" to "jul", "we'll" to "wil",
        "i've" to "aɪv", "you've" to "juv", "we've" to "wiv",
        // commands a voice bot hears a lot
        "turn" to "tɝn", "set" to "sɛt", "show" to "ʃoʊ",
        "tell" to "tɛl", "play" to "pleɪ", "pause" to "pɔz",
        "next" to "nɛkst", "back" to "bæk", "off" to "ɔf", "on" to "ɔn"
    )
}
